### PR TITLE
Desarrollo Ramo 1

### PR DESCRIPTION
**Pull request by RamonCMo**
# Module_1

## SRC

En main() se aplican los pasos para conseguir los gráficos deseados. Las calls a la API se hacen en un iterable de las ciudades pedidas (y se esperan 10s entre cada call, por seguridad, pero no es necesario esperar tanto).

La solución propuesta consta de tres partes: [get_data_meteo_api, data_parse, graficar]

### 1.-Get data

Se formatea la URL de la API para que sea flexible a distintas ciudades (lat/lon), variables, y periodos. Los errores que pueden surgir al hacer la call de la api se gestionan con la funcion call_api(). Ademas con el paquete Backoff, en caso de error por time-limit, el programa se queda en espera y vuelve a intentar la call cuando haya pasado el cold down period. Además se comprueba el schema del json que da la api con la librería jsonschema y el JSON 'schema.json' del directorio. Esta funcion devuelve los datos 'Dailies'

### 2.-Data Parse

Una vez tenemos la raw data aplicamos la funcion data_parse(), que hace la media de los distintos modelos climaticos por variable y resulta un df con 4 columnas (para la ciudad i-essima):

|          time       | temperature_2m_mean | precipitation_sum        | soil_moisture_0_to_10cm_mean |    city   |
| --------------- | ------------------------- | ------------------------- | ---------------------------------- | -------- |
| 2022-01-01  	 | 7.085714  	                 | 1.418571  	                | 0.2925  	                                    | Madrid |

Para poder calcular los intervalos de confianza (t-student) aplicamos la funcion conf_interval(), que pasa df['time']a datetime type, hace un unpivot de la tabla, Reduce la dimension temporal quedandonos solo con la media de cada variable/AÑO (como sale el display en la API-web) y calcula los ICs.

### 3.- Graficar

Graficamos y almacenamos en JPS los graficos de cada variable la tabla final (no he consegido dibujar los IC):

|          time       | Indicadores | Mean | Std |    LowerIC | UpperIC |
| --------------- | ------------------------- | ------------------------- | ---------------------------------- | -------- | -------- |
| 2020 | Temperature... | Madrid | 10 | 11 | 9 | 12 |
| 2020 | Soil... | Madrid | 10 | 11 | 9 | 12 |

## TEST

El test es sobre las funciones Data Parse y IC. Se ha cogido como muestra Madrid y la variable temperature_2m_mean y se han comparado los resultados de las funciones MAIN() con funciones que lo calculan manualmente.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,3 +72,12 @@ Los datos encontraréis disponibles en s3 son:
 
 La primera vez que vayais a trabajar con los datos, debeis descargarlos utilizando python `boto3` o `AWS CLI` y a partir de ahí guardarlos en local para evitar tener que descargarlos cada vez (IMPORTANTE: Se considera una buena práctica no añadir datos a vuestros repositorios de código ya que eso haría el repositorio muy pesado y todas las operaciones de version control muy lentas.)
 
+
+
+
+
+### Comments Varios.-Ramon
```
Comment (sergiorozada12) [inline README.md:79]: Aunque es cierto que esto no es un repo al uso no pasa nada, pero en general el README se utiliza para informar a la persona no habituada al repositorio qué se puede encontrar. No es buena práctica utilizarlo para notas personales. Si has usado sklearn en vez de scikit-learn, gestiónalo y refléjalo a través de poetry.

File: src/module_1/module_1_meteo_api.py [hunk 2]
```diff
@@ -1,7 +1,141 @@
...
+import matplotlib.pyplot as plt
+import backoff
+import time
+import json
+import scipy.stats as stats
+from jsonschema import validate, SchemaError
+
+
+# Carga el archivo JSON en la variable schema_validation
+with open("/home/ramon/Zrive/zrive-ds/src/module_1/schema.json", "r") as file:
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:15]: Si tienes que hardcodear variables (el path en este caso), es buena práctica ponerlas todas al principio y en mayúsculas:

`JSON_PATH = "/home/ramon/..."
`

De esta forma, todo el que lea el repo sabrá lo que está viendo, y tendrás todas las variables que puedes cambiar manualmente en el mismo sitio.

Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:15]: Si vas a referenciar rutas dentro del mismo repositorio, hazlo con rutas relativas, no absolutas:
- Por un lado estás mostrando información sobre tu estructura de archivos.
- Por otro lado, si este código lo clona otra persona va a romper en la ejecución.

File: src/module_1/module_1_meteo_api.py [hunk 3]
```diff
@@ -1,7 +1,141 @@
...
+    schema_to_validate = json.load(file)
+
+
+@backoff.on_exception(backoff.expo, requests.exceptions.Timeout)
+def call_api(url, params=None):
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:20]: Trata de usar typings en los parámetros de entrada sobre todo, y si puedes  de salida. Por ejemplo, leyendo esta función puedo inferir qué es `url`, pero ni idea de que es `params`. Los typings hacen el código más claro:

`call_api(url: str, params: Dict[str, Any])`

Esto aplica a todas las funciones del script

Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:20]: Creo que `params` no se usa en ninguna llamada a la función en todo el script. Si algo no se usa, es código que se puede quitar. De otra forma se convierte en lo que llamamos deuda técnica, código inútil potencial fuente de bugs, y confusiones.

File: src/module_1/module_1_meteo_api.py [hunk 4]
```diff
@@ -1,7 +1,141 @@
...
+        response = requests.get(url, params=params)
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.RequestException as e:
+        print(f"Error al hacer la solicitud a la API: {e}")
+        return None
+
+
+def get_data_meteo_api(coordinates, start_date, end_date, vars):
+    URL = f"https://climate-api.open-meteo.com/v1/climate?latitude={coordinates['latitude']}&longitude={coordinates['longitude']}&start_date={start_date}&end_date={end_date}&models=CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S&daily={vars}"
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:31]: Esta línea es muy poco clara:
- Considera introducir lógica para construir la URL: usar una función, partirla en varias variables y combinarlas, etc.
- De nuevo, las partes que están hardcodeadas, muévelas arriba ponlas y en mayúsculas:

`API_URL = "https://climate-api.open-meteo.com/v1/climate?"`

File: src/module_1/module_1_meteo_api.py [hunk 5]
```diff
@@ -1,7 +1,141 @@
...
+    data = call_api(URL)
+    try:
+        validate(data, schema_to_validate)
+    except SchemaError as e:
+        print(f"Error en el API_JSON schema{e}")
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:36]: Aplica a varios puntos. La idea de printear el error es buena, pero es poco robusto para un código "productivo". La forma más profesional de printear información es usar logging de python:

https://docs.python.org/3/howto/logging.html

Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:36]: Cuando hay un error, hay que gestionarlo. Por ejemplo, si llamamos varias veces a la API podemos contarlos para saber si el error es siempre el mismo. Otra opción es romper la ejecución del programa si encontramos un error con la función `raise`

File: src/module_1/module_1_meteo_api.py [hunk 6]
```diff
@@ -1,7 +1,141 @@
...
+    )
+    result["Upper_CI"] = result.apply(
+        lambda row: row["mean"]
+        + stats.t.ppf(1 - 0.05 / 2, df=len(df) - 1) * row["std"] / (len(df) ** 0.5),
+        axis=1,
+    )
+    return result
+
+
+def graficar(df, variables):
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:72]: El script está íntegramente en inglés, menos el nombre de esta función

File: src/module_1/module_1_meteo_api.py [hunk 7]
```diff
@@ -1,7 +1,141 @@
...
+        df = pd.DataFrame(data=data["daily"])
+    return df
+
+
+def data_parse(df, variables):
+    vars = [x.strip() for x in variables.split(",")]
+    resumen_df = df[["time"]].copy()
+    for variable in vars:
+        modelos = [col for col in df.columns if variable in col]
+        promedio_por_dia = df[modelos].mean(axis=1)
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:47]: Ojo con mezclar nombres de variables en inglés y en español

File: src/module_1/module_1_meteo_api.py [hunk 8]
```diff
@@ -1,7 +1,141 @@
...
+            ax = axs[i]
+        else:
+            ax = axs
+        # Configurar el gráfico
+        ax.set_title(variable)  # Establecer el título del gráfico como la variable
+        for city in cities:
+            city_data = df_filtered[df_filtered["city"] == city]
+            # Dibujar el gráfico de líneas
+            ax.plot(city_data["Year"], city_data["mean"], label=city)
+            # Colorear el intervalo de confianza
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:93]: Este tipo de comentarios no son necesarios. Hay varios similares. El comportamiento de la función se infiere del nombre. Los comentarios solo se ponen cuando añaden claridad.

File: src/module_1/module_1_meteo_api.py [hunk 9]
```diff
@@ -1,7 +1,141 @@
...
+        ax.legend(loc="lower center")
+        ax.set_title(f"{variable} with Confidence Intervals by City")
+    plt.tight_layout()
+    fig.savefig("Meteo.jpg", format="jpg")
+
 
 def main():
-    raise NotImplementedError
+    # Variables iniciales
+    COORDINATES = {
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:112]: De nuevo, mover todas estas variables hardcodeadas al inicio del script

File: src/module_1/module_1_meteo_api.py (hunk 3 above)
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:20]: Una buena práctica cuando llamamos APIs es hacerlo varias veces. A veces las llamadas fallan por cuestiones externas, y necesitamos repetirlas.

File: src/module_1/module_1_meteo_api.py [hunk 10]
```diff
@@ -1,7 +1,141 @@
...
+        variable_intervals = conf_interval(df=viariable_data)
+        variable_intervals["city"] = i
+        data_cities.append(variable_intervals)
+        time.sleep(3)
+    meteo_data = pd.concat(data_cities, axis=0)
+    graficar(df=meteo_data, variables=VARIABLES)
+
 
 if __name__ == "__main__":
+    start_time = time.time()
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:138]: Puedes mover la lógica del timer dentro de la función `main`

File: src/module_1/module_1_meteo_api.py [hunk 11]
```diff
@@ -1,7 +1,141 @@
...
+        # Configurar el gráfico
+        ax.set_title(variable)  # Establecer el título del gráfico como la variable
+        for city in cities:
+            city_data = df_filtered[df_filtered["city"] == city]
+            # Dibujar el gráfico de líneas
+            ax.plot(city_data["Year"], city_data["mean"], label=city)
+            # Colorear el intervalo de confianza
+            ax.fill_between(
+                city_data["Year"],
+                city_data["Lower_CI"],
```
Comment (sergiorozada12) [inline src/module_1/module_1_meteo_api.py:96]: Los intervalos de confianza no se ven en la imagen `Meteo.jpg`.

File: tests/module_1/test_meteo_api.py [hunk 12]
```diff
@@ -1,7 +1,40 @@
...
+    city_data = get_data_meteo_api(
+        coordinates=COORDINATES['Madrid'],
+        start_date=start_date,
+        end_date=end_date,
+        vars=VARIABLES,
+    )
+    viariable_data = data_parse(df=city_data, variables=VARIABLES)
+    test_mean = mean_test(city_data)
+    variable_intervals = conf_interval(df=viariable_data)
+    print(viariable_data.head())
```
Comment (sergiorozada12) [inline tests/module_1/test_meteo_api.py:32]: Los tests no necesitan este tipo de prints. No es un entorno de desarrollo. Estás printeando una entrada, y ya sabes lo que va a ser. Los tests son totalmente funcionales, para comprobar que una función hace lo que quieres. Si quieres meter algún log de inicio, o de que has acabado, o que vas completando los tests está bien.

File: tests/module_1/test_meteo_api.py [hunk 13]
```diff
@@ -1,7 +1,40 @@
...
+    print(variable_intervals.head())
+    print(test_mean)
+
+if __name__ == "__main__":
+    start_time = time.time()
```
Comment (sergiorozada12) [inline tests/module_1/test_meteo_api.py:37]: En los tests no medimos latencias, o tiempo de ejecución. El propósito del test es comprobar que el código hace aquello para lo que está diseñado.

File: tests/module_1/test_meteo_api.py [hunk 14]
```diff
@@ -1,7 +1,40 @@
 """ This is a dummy example to show how to import code from src/ for testing"""
+import os
+import sys
+import time
+sys.path.append(os.path.join(os.path.dirname(__file__),'../../'))
+from src.module_1.module_1_meteo_api import call_api
+from src.module_1.module_1_meteo_api import get_data_meteo_api
+from src.module_1.module_1_meteo_api import data_parse
+from src.module_1.module_1_meteo_api import conf_interval
```
Comment (sergiorozada12) [inline tests/module_1/test_meteo_api.py:9]: La idea de testear está bien, pero no hace falta que hagas todo desde cero. En python hay librerías de testeo que hace toda la parte complicado por nosotros. Échale un ojo a pytest.

File: tests/module_1/test_meteo_api.py [hunk 15]
```diff
@@ -1,7 +1,40 @@
...
 
-from src.module_1.module_1_meteo_api import main
-
+def mean_test(df):
```
Comment (sergiorozada12) [inline tests/module_1/test_meteo_api.py:11]: La idea está bien, pero faltan tests. Con los tests buscamos comprobar funcionalmente lo que hace nuestro código, especialmente en los corner cases:
-  ¿Qué pasa si la API falla?
-  ¿Qué pasa si el schema no está bien?
-  ¿Qué pasa si la API da un timeout?
-  ¿Qué pasa si la API no falla pero da una respuesta vacía?

**Review by sergiorozada12 — APPROVED** [review]
Buen trabajo, sigue dándole caña. Te apruebo para que mergees. No obstante, échale un vistazo a los comentarios que te he dejado. La idea es que los tengas en cuenta para futuras PRs.

Un comentario general con respecto al nombre de la rama. Las ramas son features, cambios que añaden funcionalidad al código, o fixes, arreglos, cambios menores. El nombre tiene que reflejarlo, y `desarrollo_ramon_1` no dice mucho. Un nombre alternativo (más claro) sería `feature/meteo_api`, por ejemplo.

**Comment by RamonCMo** [comment]
Muy claros los comentarios! A veces se me olvida que no solo escribo para mi 😅. Seguiré mirando la lógica de como hacer los test para la próxima que creo que es lo que peor llevo. Ty Sergio

**Comment by sergiorozada12** [comment]
Para eso estamos, cualquier cosa con la que te pueda ayudar me dices sin problema 🙂
